Dlipko

dnlipko opened the pull request:
1. Обмен данными между сервером и клиентом происходит через xml файлы

- Клиент подключается к серверу и отправляет xml с логином и паролем
(<user><signup><login>abcd</login><password>12345</password></signup></user>)

- Клиент отправляет на сервер команду о желании начать партию
(<user><start>1</start></user>)

-Клиент получает данные: победа, количество побед в раунде,  его ли очередь хода, карты в руке, карты на поле, количество карт в руке соперника

2. Об игре
- Игра длиться 2 или 3 раунда в зависимости от того кто победит в 1 и 2 раунде. В игре побеждает тот кто быстрее наберет 2 победы в раундах.  3ий раунд играется, если через 2 раунда ничья.
- В раунде побеждает тот кто наберет больше всего очков.
- Очки складываются из силы карт лежащих на поле 
- Раунд завершается когда пасанут оба игрока.
-После того как пасанет один игрок, другой может продолжать кидать карты пока он сам не пасанет.
-За ход можно сыграть одну карту из руки (не считая карт со способностями, которые могут вытягивать на поле другие карты)
-В каждом раунде поле обновляется(не считая карт со специальных карт со способностями)
-Колода одна на всю игру
-Карты в руке и колода при перехода от раунда к раунду не изменяются 

3. Карты(будет дополняться, тестовая версия)
Есть три типа карт:
-Сущность (обладает силой)
Скорее всего от сущности будет зависеть её сила
-Сущность со способностью(обладает силой и способностью)
-Способность (Какие-нибудь карты погоды или что-нибудь подобное)
3.1. buff Удваивает свою силу и силу карты на поле с соответствующей сущностью (Для сущности со способностью)
3.2. spy карта кладется на стол противника, но в руку получаете +2 карты (Для сущности со способностью)
3.3 rollcall вытягивает подобную сущность из колоды вместе с собой на поле (Для сущности со способностью)

Воспользовался паттерном Factory Method. Но генерация карт пока не продуманная. Планирую вынести генерацию карт в отдельный класс. Пока что все в юзере.
Перенес генерацию xml(qdomelement) в сами классы.

dnlipko committed:
Server first steps

Есть возможность подключится к серверу.
Он объединяет в парочки каждых двух подключившихся.
Потом если один из них пишет на сревер, он отправляет это обратно отправителю и его противнику по игре.

dnlipko committed:
user ligin

user may login or signep by sending xml
<user><signup><login>dlipko</login><password>12345</password></signup></user>

or

<user><login><login>dlipko</login><password>12345</password></login></user>

dnlipko committed:
рабочий вариант

Добавил коментарии. Генерации карт нет. Из карт пока только студет. Свойств у карт пока тоже нет.

dnlipko committed:
Factory Method, qdom

Добавил карты сущности со свойствами, просто сущности и карты со свойствами.  Воспользовался паттерном Factory Method. Но генерация карт пока не продуманная. Планирую вынести генерацию карт в отдельный класс. Пока что все в юзере.
Перенес генерацию xml(qdomelement) в сами классы.

dnlipko committed:
delete this sh**

dnlipko committed:
client

Накидал пару классов на клиенте, разделил логигу сервера промежуточного слогя и игры.
Планирую парсить данные и создавать объекты // сравнивать их с существующими. И уже слой игровой логики прикрепить к GUI

dnlipko committed:
Merge branch 'dlipko' of https://github.com/lostmap/CardGame into dlipko

leshiy1295 commented on client/abstractcard.cpp:
```diff
@@ -0,0 +1,17 @@
+#include "abstractcard.h"
+
+AbstractCard::AbstractCard(int id, QString info):
```
отделите код от Qt, где в этом нет необходимости

leshiy1295 commented on client/abstractcard.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef ABSTRACTCARD_H
+#define ABSTRACTCARD_H
+
+#include <QString>
```
Qt - это реализация, поэтому постарайтесь не включать без необходимости его библиотеки в .h файлах

leshiy1295 commented on client/client.cpp:
```diff
@@ -0,0 +1,49 @@
+{
+    _socket = new QTcpSocket(this);
+
+    _socket->connectToHost("127.0.0.1", 1234);
```
это необходимо пробрасывать через конструктор и по-хорошему брать либо из argv, либо из конфигурационных файлов

leshiy1295 commented on client/client.cpp:
```diff
@@ -0,0 +1,49 @@
+
+void Client::disconnected()
+{
+    qDebug() << "disconnected...";
```
лучше создать отдельную сущность логгера и пользоваться ей через единый интерфейс везде, в main же создавать реализацию на Qt

leshiy1295 commented on client/client.cpp:
```diff
@@ -0,0 +1,49 @@
+
+    QByteArray data;
+    data = _socket->readAll();
+    while (!data.contains("</server>") && _socket->waitForReadyRead()) {
```
кажется, что это хардкод

leshiy1295 commented on client/client.cpp:
```diff
@@ -0,0 +1,49 @@
+    _interLayer->parseData(data);
+}
+
+void Client::send(QString data)
```
это же view-класс, поэтому стоит вынести модельную логику из него, исходя из принципа единственной ответственности

leshiy1295 commented on client/client.h:
```diff
@@ -0,0 +1,35 @@
+
+#include <QTcpSocket>
+#include <QAbstractSocket>
+#include <QDebug>
```
этого нет в интерфейсе, поэтому и подключать нужно в файле реализации

leshiy1295 commented on client/deck.cpp:
```diff
@@ -0,0 +1,42 @@
+
+AbstractCard * Deck::findByCardId(int cardId)
+{
+    for (QVector<AbstractCard *>::iterator i = _deck.begin();
```
find_if...

leshiy1295 commented on client/deck.cpp:
```diff
@@ -0,0 +1,42 @@
+
+AbstractCard* Deck::removeCardByCardId(int cardId)
+{
+    AbstractCard * card = findByCardId(cardId);
```
аналогично - разнесите интерфейсный класс и модельный

leshiy1295 commented on client/deck.cpp.autosave:
```diff
@@ -0,0 +1,58 @@
+#include "deck.h"
```
кажется, что этот файл лишний

leshiy1295 commented on client/deck.h:
```diff
@@ -0,0 +1,30 @@
+    int size();
+
+private:
+    QVector<AbstractCard *> _deck;
```
избавьтесь от Qt в интерфейсе
в некоторых случаях - можно воспользоваться идиомой PIMPL

leshiy1295 commented on client/entitycard.cpp:
```diff
@@ -0,0 +1,29 @@
+
+Property_type EntityCard::getPropertyType()
+{
+    return NO_PROPERTI;
```
PROPERTY..

leshiy1295 commented on client/entitycard.h:
```diff
@@ -0,0 +1,22 @@
+class EntityCard: virtual public AbstractCard
+{
+public:
+    EntityCard(int id, QString info, Entity_type entity_type, int strength);
```
убрать QString

leshiy1295 commented on client/entitywithpropertycard.h:
```diff
@@ -0,0 +1,21 @@
+class EntityWithPropertyCard: public EntityCard, public PropertyCard
+{
+public:
+    EntityWithPropertyCard(int id, QString info, Entity_type entity_type,
```
убрать Qt

leshiy1295 commented on client/interlayer.cpp:
```diff
@@ -0,0 +1,141 @@
+{
+    QTime midnight(0,0,0);
+    qsrand(midnight.secsTo(QTime::currentTime()));
+    sendLogin(QString("dlipko" + QString::number(qrand()%1000)), QString::number(12345));
```
хардкод
и это скорее всего должна быть отдельная структура, которая умеет себя сериализовывать в строку

leshiy1295 commented on client/interlayer.cpp:
```diff
@@ -0,0 +1,141 @@
+                    QDomNode meNode = node.firstChild();
+
+                    while (!meNode.isNull()){
+                        if (meNode.toElement().tagName() == "me"){
```
стоит вынести в ENUM

leshiy1295 commented on client/interlayer.cpp:
```diff
@@ -0,0 +1,141 @@
+                }
+
+
+                if (command == "heand")
```
нет опечатки?

leshiy1295 commented on client/interlayer.cpp:
```diff
@@ -0,0 +1,141 @@
+
+                            while (!meNodeChild.isNull()){
+                                QString command = meNodeChild.toElement().tagName();
+                                if (command == "stagescore"||command == "score"
```
лучше навесить фильтров - отдельных классов, которые внутри себя через единый интерфейс проверяют Ваш объект на соответствие той или иной команды.
Соответственно, проход по фильтрам завершается, когда кто-то вернёт true, то есть что он принял запрос на обработку

leshiy1295 commented on client/interlayer.cpp:
```diff
@@ -0,0 +1,141 @@
+
+void InterLayer::sendCard(int id)
+{
+    _client->send(QString("<user><card><id>" + QString::number(id) + "</id></card></user>").toUtf8());
```
это должен быть метод сериализации клиента

leshiy1295 commented on client/interlayer.h:
```diff
@@ -0,0 +1,33 @@
+    MainWindow *_mainWindow;
+    Client *_client;
+
+    void _parseNode(QDomElement);
```
вынесите в отдельный интерфейс, который не зависит от Qt

leshiy1295 commented on client/main.cpp:
```diff
@@ -0,0 +1,22 @@
+
+int main(int argc, char *argv[])
+{
+    QApplication a(argc, argv);
```
лучше также создать отдельный абстрактный класс Application, в котором всё инициализировать, в том числе и Qt в одной из реализаций

leshiy1295 commented on client/mainwindow.cpp:
```diff
@@ -0,0 +1,60 @@
+
+MainWindow::~MainWindow()
+{
+    delete ui;
```
можно использовать умный указатель

leshiy1295 commented on client/mainwindow.h:
```diff
@@ -0,0 +1,42 @@
+private:
+        Ui::MainWindow *ui;
+
+        // пока что добавлю сюда
```
нужно доделать

leshiy1295 commented on client/my_qt_chat_client.pro.user:
```diff
@@ -0,0 +1,318 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
видимо, временный файл

leshiy1295 commented on client/user.cpp.autosave:
```diff
@@ -0,0 +1,7 @@
+#include "user.h"
```
все ненужные файлы нужно удалить и добавить в gitignore

leshiy1295 commented on server/abstractcard.cpp:
```diff
@@ -0,0 +1,53 @@
+    return _id;
+}
+
+QDomElement AbstractCard::toDomElement()
```
сейчас есть привязка к Qt
нужно добавить класс-развязку, который будет адаптером между этим классом и сериализующей в Qt реализацией

leshiy1295 commented on server/cardfactory.cpp:
```diff
@@ -0,0 +1,8 @@
+#include "cardfactory.h"
+
+int CardFactory::cards = 0;
```
?

leshiy1295 commented on server/deck.cpp:
```diff
@@ -0,0 +1,84 @@
+
+AbstractCard * Deck::findByCardId(int cardId)
+{
+    for (QVector<AbstractCard *>::iterator i = _deck.begin();
```
find_if
похожий код был выше

leshiy1295 commented on server/deck.cpp:
```diff
@@ -0,0 +1,84 @@
+
+AbstractCard* Deck::findByEntityType(int entityType)
+{
+    for (QVector<AbstractCard *>::iterator i = _deck.begin();
```
find_if

leshiy1295 commented on server/deck.cpp:
```diff
@@ -0,0 +1,84 @@
+    QDomDocument document;
+    QDomElement element = document.createElement(deckName);
+
+    for (auto i = _deck.begin(); i != _deck.end(); ++i)
```
используйте for-range циклы

leshiy1295 commented on server/define.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef DEFINE_H
+#define DEFINE_H
+
+#define HEAND_SIZE 10
```
это стоит вынести в конфигурационный файл, чтобы не перекомпилировать весь проект

leshiy1295 commented on server/define.h:
```diff
@@ -0,0 +1,13 @@
+
+enum Property_type {NO_PROPERTI, Buff, Spy, RollCall};
+
+enum Entity_type {NO_ENTITY, Dragon, Woodcutter};
```
лучше использовать единый стиль
константы обычно пишутся большими через _

leshiy1295 commented on server/entity.cpp:
```diff
@@ -0,0 +1,31 @@
+#include "entity.h"
```
похожий класс был выше

leshiy1295 commented on server/entityfactory.cpp:
```diff
@@ -0,0 +1,12 @@
+
+AbstractCard *EntityFactory::createCard(QString info, Entity_type entity_type, int strength)
+{
+    return new Entity(cards++, info, entity_type, strength);
```
используйте умные указатели
например, тут стоит использовать unique_ptr

leshiy1295 commented on server/entitywithbuff.cpp:
```diff
@@ -0,0 +1,23 @@
+        AbstractCard *same_card = deck->findByEntityType(getEntityType());
+        if (same_card)
+        {
+            same_card->setStrength(same_card->getStrength()*2);
```
2 - магическая константа
нужно вынести в константу [и конфиг]

leshiy1295 commented on server/entityrollcallfactory.cpp:
```diff
@@ -0,0 +1,12 @@
+
+AbstractCard* EntityRollCallFactory::createCard(QString info, Entity_type entity_type, int strength)
+{
+    return new EntityRollCall(cards++, info, entity_type, strength);
```
и тут

leshiy1295 commented on server/field.cpp:
```diff
@@ -0,0 +1,69 @@
+
+Field::~Field()
+{
+    for ( QMap<Player *, Deck *>::iterator deck = _tableDeck.begin(); deck != _tableDeck.end(); ++deck)
```
auto for range

leshiy1295 commented on server/field.cpp:
```diff
@@ -0,0 +1,69 @@
+void Field::addCardToField(Player *player1, Player *player2, AbstractCard *card)
+{
+    switch (card->property()) {
+    case Buff:
```
используйте делегирование (например, полиморфное - см. последнюю лекцию)
сейчас получается хардкод в одном классе

leshiy1295 commented on server/field.cpp:
```diff
@@ -0,0 +1,69 @@
+        break;
+
+    case RollCall:
+        // !!!!!!!!!!!!!!!!!!!!!
```
?

leshiy1295 commented on server/field.h:
```diff
@@ -0,0 +1,35 @@
+#include "player.h"
+#include "abstractcard.h"
+
+#include <QMap>
```
вынести Qt из интерфейса

leshiy1295 commented on server/game.cpp:
```diff
@@ -0,0 +1,155 @@
+{
+    for (QMap<QString, User *>::iterator user = _user.begin(); user != _user.end(); ++user)
+    {
+        delete (*user);
```
используйте по возможности умные указатели для модельных сущностей

leshiy1295 commented on server/game.cpp:
```diff
@@ -0,0 +1,155 @@
+        Player *player2 = new Player(_getUser(login));
+
+        // объединяет их в пару
+        _player[_getUser(_waitingForCouple)] = player1;
```
что будет, если _getUser вернёт невалидное значение?

leshiy1295 commented on server/game.cpp:
```diff
@@ -0,0 +1,155 @@
+
+Player* Game::_getPlayerByLogin(QString login)
+{
+    return _player[_user[login]];
```
аналогично - что будет, если _user[login] вернёт невалидное значение...?

leshiy1295 commented on server/game.h:
```diff
@@ -0,0 +1,66 @@
+#include <QDomDocument>
+
+
+class Game
```
сильно привязанный к Qt класс

leshiy1295 commented on server/interlayer.cpp:
```diff
@@ -0,0 +1,109 @@
+
+InterLayer::~InterLayer()
+{
+    if (_game)
```
используйте умные указатели

leshiy1295 commented on server/interlayer.cpp:
```diff
@@ -0,0 +1,109 @@
+
+        qDebug() << command;
+
+        if (command == "login" or command == "signup")
```
похожий класс был выше
на сервере и клиенте стоит добиться создание идемпотентного кода, то есть одинакового для его прямого переиспользования

leshiy1295 commented on server/interlayer.h:
```diff
@@ -0,0 +1,48 @@
+    // соответсвие логина и сокета
+    QMap<QString, QTcpSocket *> _socket;
+    // соответствие сокета и логина
+    QMap<QTcpSocket *, QString> _login;
```
скройте QTcpSocket за абстракцией сокета с нужным Вам интерфейсом

leshiy1295 commented on server/my_chat_qt_server.pro.user:
```diff
@@ -0,0 +1,318 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
ещё один лишний файл

leshiy1295 commented on server/myserver.cpp:
```diff
@@ -0,0 +1,71 @@
+
+void MyServer::disconnected()
+{
+    /*
```
?

leshiy1295 commented on server/myserver.cpp:
```diff
@@ -0,0 +1,71 @@
+
+    QByteArray data;
+    data = client->readAll();
+    while (!data.contains("</user>") && client->waitForReadyRead()) {
```
хардкод

leshiy1295 commented on server/player.cpp:
```diff
@@ -0,0 +1,206 @@
+#include <QTime>
+#include <QDebug>
+
+#include "entitywithbuff.h"
```
не очень удачные названия классов

leshiy1295 commented on server/player.cpp:
```diff
@@ -0,0 +1,206 @@
+    qsrand(midnight.secsTo(QTime::currentTime()));
+
+
+    for (int i= HEAND_SIZE; i < 4; ++i)
```
хардкод

leshiy1295 commented on server/player.cpp:
```diff
@@ -0,0 +1,206 @@
+        newDeck->addCard(new EntityWithBuff(i, "Dragon with couple buff", Dragon, qrand() % MAX_STRENGTH + 1));
+    }
+
+    for (int i= 4; i < DECK_SIZE; ++i)
```
это должно быть в конфигурационных файлах и считываться при инициализации приложения при создании фабрики возможных существующих карт

leshiy1295 commented on server/player.cpp:
```diff
@@ -0,0 +1,206 @@
+
+    for (int i= 4; i < HEAND_SIZE; ++i)
+    {
+        newDeck->addCard(new EntitySpy(i, "Woodcutter Spy", Woodcutter, qrand() % MAX_STRENGTH + 1));
```
не нужно использовать Qt без необходимости

leshiy1295 commented on server/player.cpp:
```diff
@@ -0,0 +1,206 @@
+
+void Player::setWin()
+{
+    _win = 1;
```
хардкод

leshiy1295 commented on server/player.cpp:
```diff
@@ -0,0 +1,206 @@
+
+QDomElement Player::toQDomElement(QString playerName)
+{
+    QDomDocument document;
```
методы сериализации стоит вынести в отдельный интерфейс, от которого наследовать Player

leshiy1295 commented on server/player.h:
```diff
@@ -0,0 +1,73 @@
+    Deck *_deckGenerator();
+    Deck *_heandGenerator();
+
+    QDomElement _domElement(QString elementName, int value);
```
вынести Qt из реализации

leshiy1295 commented on server/user.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef USER_H
```
?

leshiy1295 reviewed (changes requested):
сейчас самая большая проблема - большая привязка к Qt

dnlipko committed:
+конфиг, логгер, pugixml. -qt

Полностью избавился от привязки к qt там где в этом нет необходимости. Добавил конфигурационный файл и логгер. Постарался воспользоваться умными указателями.  По возможности исправил все указанные недочеты.

dnlipko committed:
delete lod file

leshiy1295 commented on server/configparser.h:
```diff
@@ -0,0 +1,33 @@
+    std::map<PROPERTY_TYPE, std::string> getPropertiesInfo() const;
+
+private:
+  pugi::xml_document doc;
```
зависимость осталась в header-файле - нехорошо
в качестве решения можно предложить использовать идиому PIMPL

leshiy1295 commented on server/game.cpp:
```diff
@@ -0,0 +1,212 @@
+        if (!_waitingForCouple.empty()){
+
+            // создает игроков
+            auto player1 = std::shared_ptr<Player>(new Player(_getUser(_waitingForCouple)));
```
есть make_shared

leshiy1295 commented on server/interlayer.cpp:
```diff
@@ -0,0 +1,170 @@
+InterLayer::~InterLayer() {
+
+    for (auto socket: _socket)
+        delete socket.second;
```
удобнее использовать умные указатели

dnlipko replied:
Уже несколько раз пытался использовать с сокетами shared_ptr. Каждый раз появляется не понятная проблема, с которой мне пока что так и не удалось разобраться.

leshiy1295 commented on server/interlayer.h:
```diff
@@ -0,0 +1,60 @@
+    std::string _parseGetStr3(pugi::xml_document &message,std::string ancestor1,
+                              std::string ancestor2, std::string parent) const;
+
+    bool _signIn(pugi::xml_document &message, Socket* client);
```
PIMPL

leshiy1295 commented on server/socket.h:
```diff
@@ -0,0 +1,17 @@
+
+#include <string>
+
+class Socket: public QTcpSocket
```
нужно сделать Socket абстрактным
и QSocket уже Ваш, который будет наследоваться от Socket и от QTcpSocket

leshiy1295 commented on server/game.h:
```diff
@@ -0,0 +1,70 @@
+#include <memory>
+
+
+class Game: QDomSerializer
```
немного не так - Вы сделали Game QT-классом
нужно сделать абстрактный интерфейс Serializer, добавить поле Parser внутрь Game, передавать ему QtParser, который уже будет сериализовывать

leshiy1295 reviewed (changes requested):
Намного лучше
Нужно немного доправить архитектуру

dnlipko committed:
PIMPL, QSocket, Serializer

Использовал идиому PIMPL для pugi.
Сделал Socket абстрактным классом. Создал класс QSocket, который унаследовал от Socket и от QTcpSocket.
Сделать абстрактный интерфейс Serializer, добавил поле Parser внутрь Game, в который передаю QDomSerializer.

dnlipko committed:
Merge branch 'dlipko' of https://github.com/lostmap/CardGame into dlipko

leshiy1295 commented on server/configparser.h:
```diff
@@ -25,7 +28,7 @@ class ConfigParser
 
 private:
-  pugi::xml_document doc;
+    std::shared_ptr<pugi::xml_document> doc;
```
это не PIMPL..
PIMPL - это когда тут Вы оставляете указатель на класс реализации...
А в нём уже подключаете этот файл.
Из этого же класса Вы проксируете все запросы в класс реализации

leshiy1295 commented on server/game.cpp:
```diff
@@ -40,8 +40,8 @@ void Game::findCouple(std::string login)
             // создает игроков
-            auto player1 = std::shared_ptr<Player>(new Player(_getUser(_waitingForCouple)));
-            auto player2 = std::shared_ptr<Player>(new Player(_getUser(login)));
+            auto player1 = std::make_shared<Player>(_getUser(_waitingForCouple));
```
когда тип однозначно определён, то make_shared можно явно не инстанцировать - в этом вся соль)

leshiy1295 commented on server/server.h:
```diff
@@ -1,31 +1,24 @@
+#include <QHostAddress>
 
-class Server: public QTcpServer
+class Server
```
раз это сервер, который не привязан к Server, то у него не должно быть методов с QByteArray - нужно просто с байтами. В Вашем же классе Вы просто преобразуете QByte к обычным (ну или наоборот)
